src/Tools/Microsoft.dotnet-openapi/src/Commands/BaseCommand.cs

```diff
@@ -0,0 +1,524 @@
+            }
+            else
+            {
+                WorkingDirectory = ((Application)Parent.Parent).WorkingDirectory;
+            }
+
+            OnExecute(ExecuteAsync);
+        }
+
+        public CommandOption ProjectFileOption { get; }
+
+        internal CommandOption Help { get; }
+
+        public TextWriter Warning
+        {
+            get { return Out; }
+        }
+
+        protected abstract Task<int> ExecuteCoreAsync();
+
+        protected abstract bool ValidateArguments();
+
+        private async Task<int> ExecuteAsync()
+        {
+            if (!ValidateArguments() || Help.HasValue())
+            {
+                ShowHelp();
+                return 1;
+            }
+
+            return await ExecuteCoreAsync();
+        }
+
+        internal FileInfo ResolveProjectFile(CommandOption projectOption)
+        {
+            string project;
+            if (projectOption.HasValue())
+            {
+                project = projectOption.Value();
+                project = GetFullPath(project);
+                if (!File.Exists(project))
+                {
+                    throw new ArgumentException($"The project '{project}' does not exist.");
+                }
+            }
+            else
+            {
+                var projects = Directory.GetFiles(WorkingDirectory, "*.csproj", SearchOption.TopDirectoryOnly);
+                if (projects.Length == 0)
+                {
+                    throw new ArgumentException("No project files were found in the current directory. Either move to a new directory or provide the project explicitly");
+                }
+                if (projects.Length > 1)
+                {
+                    throw new ArgumentException("More than one project was found in this directory, either remove a duplicate or explicitly provide the project.");
+                }
+
+                project = projects[0];
+            }
+
+            return new FileInfo(project);
+        }
+
+        protected Project LoadProject(FileInfo projectFile)
+        {
+            var project = ProjectCollection.GlobalProjectCollection.LoadProject(
+                projectFile.FullName,
+                globalProperties: null,
+                toolsVersion: null);
+            project.ReevaluateIfNecessary();
+            return project;
+        }
+
+        internal bool IsProjectFile(string file)
+        {
+            return File.Exists(Path.GetFullPath(file)) && file.EndsWith(".csproj");
+        }
+
+        internal bool IsUrl(string file)
+        {
+            return Uri.TryCreate(file, UriKind.Absolute, out var _) && file.StartsWith("http");
+        }
+
+        internal async Task AddOpenAPIReference(
+            string tagName,
+            FileInfo projectFile,
+            string sourceFile,
+            CodeGenerator? codeGenerator,
+            string sourceUrl = null)
+        {
+            await EnsurePackagesInProjectAsync(projectFile, codeGenerator);
```

Comment:
Couple reasons:
1. We have to `EnsurePackagesInProjectAsync` for each codeGenerator because even though they both use the same packages now, future instances will almost certainly have different dependencies. I know the current command does multiple files for one code generator, but if we revisit this in the future I'd like to decouple that, and this format makes it future-proof against future-Ryan who has forgotten this future-fact.
1. LoadProject uses `ProjectCollection`, which is super useful but for some reason creates global state, which it's hard to mark as invalid, so it's very important that the package edits (which don't use ProjectCollection) happen BEFORE the OpenApiReference edits, because if not the package edits tend to get lost (because I couldn't find a way to force Project to re-load the csproj).
1. Testing. The aforementioned global state makes testing these without spinning up multiple processes a pain.

This pattern is the best one I found to handle all three problems.